core/nodefactory.lua

```diff
@@ -261,6 +261,18 @@ end
 
 function nodefactory.unshaped:shape ()
   local node =  SILE.shaper:createNnodes(self.text, self.options)
+  if self.markedForSpeakerChange then
```

Comment:
If going this route (space handling after shaping), rather than using a mark flag and then a dumb nil check in almost everything we shape, subclassing the node would certainly be better :)

Reply:
Done. "Avoid messing with standard node classes" should be a motto.